File: ruby/proverb/proverb.rb
class Proverb

  attr_reader :items, :qualifier
  def initialize(*items, qualifier: nil)
    @items = items
    @qualifier = "#{qualifier} #{items.first}".strip
  end

  def to_s
    output = String.new
    items.each_cons(2) do |item1, item2|
      output << partial(item1, item2)
    end
    output << ending
  end

  private

  def partial(item1, item2)
    "For want of a #{item1} the #{item2} was lost.\n" 
  end

  def ending
    "And all for the want of a #{qualifier}."
  end

end

File: go/proverb/proverb.go
package proverb

import "fmt"

func Proverb(rhyme []string) []string {
	if len(rhyme) == 0 {
		return []string{}
	} else if len(rhyme) == 1 {
		return []string{footer(rhyme[0])}
	}
	return multilineProverb(rhyme)
}

func multilineProverb(rhyme []string) (result []string) {
	for i, item := range rhyme {
		if i == len(rhyme)-1 {
			result = append(result, footer(rhyme[0]))
		} else {
			result = append(result, verse(item, rhyme[i+1]))
		}
	}
	return result
}

func verse(itemA string, itemB string) string {
	return fmt.Sprintf("For want of a %s the %s was lost.", itemA, itemB)

}

func footer(item string) string {
	return fmt.Sprintf("And all for the want of a %s.", item)
}
